presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java

```diff
@@ -59,65 +88,149 @@ protected Void visitJoin(Join node, MaterializedViewPlanValidatorContext context
+        if (analyzeColumnMapping) {
+            analyzeSetOperatorForLinkedBaseColumns(node);
         }
 
-        // TODO: It should only support equi join case https://github.com/prestodb/presto/issues/16033
-        if (!node.getOperator().equals(LogicalBinaryExpression.Operator.AND)) {
-            throw new SemanticException(NOT_SUPPORTED, node, "Only AND operator is supported for join criteria for materialized view.");
+        return super.visitSetOperation(node, context);
+    }
+
+    @Override
+    protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, MaterializedViewPlanValidatorContext context)
+    {
+        if (context.isProcessingJoinNode()) {
+            if (!node.getOperator().equals(LogicalBinaryExpression.Operator.AND)) {
+                throw new SemanticException(NOT_SUPPORTED, node, "Only AND operator is supported for join criteria for materialized view.");
+            }
         }
+
         return super.visitLogicalBinaryExpression(node, context);
     }
 
     @Override
     protected Void visitComparisonExpression(ComparisonExpression node, MaterializedViewPlanValidatorContext context)
     {
-        if (!context.isProcessingJoinNode()) {
-            return super.visitComparisonExpression(node, context);
+        if (context.isProcessingJoinNode()) {
+            if (!node.getOperator().equals(ComparisonExpression.Operator.EQUAL)) {
+                throw new SemanticException(NOT_SUPPORTED, node, "Only EQUAL join is supported for materialized view.");
+            }
+
+            if (analyzeColumnMapping) {
+                analyzeComparisonExpressionForLinkedBaseColumns(node, context);
+            }
+        }
+
+        return super.visitComparisonExpression(node, context);
+    }
+
+    private void analyzeSetOperatorForLinkedBaseColumns(SetOperation node)
+    {
+        List<RelationType> outputDescriptorList = node.getRelations().stream()
+                .map(relation -> analysis.get().getOutputDescriptor(relation).withOnlyVisibleFields())
+                .collect(toImmutableList());
+
+        int numRelations = outputDescriptorList.size();
+        int numFields = outputDescriptorList.get(0).getVisibleFieldCount();
+        for (int f = 0; f < numFields; f++) {
+            for (int r = 0; r < numRelations; r++) {
+                Optional<TableColumn> firstBaseColumn = tryGetOriginalTableColumn(outputDescriptorList.get(r).getFieldByIndex(f));
+                for (int t = r + 1; t < numRelations; t++) {
+                    Optional<TableColumn> secondBaseColumn = tryGetOriginalTableColumn(outputDescriptorList.get(t).getFieldByIndex(f));
+                    if (firstBaseColumn.isPresent() && secondBaseColumn.isPresent() && !firstBaseColumn.get().equals(secondBaseColumn.get())) {
+                        linkedBaseColumns.add(ImmutableList.of(firstBaseColumn.get(), secondBaseColumn.get()));
+                    }
+                }
+            }
         }
```

Comment:
Haven't looked into details of this part. There are some naming convention issue. Let's not use `num`, `f`, or `r`. We don't know what they stand for.

Reply:
Renamed